src/ast/pass_manager.h

```diff
@@ -1,64 +1,366 @@
+  // that are references, this should be generally discouraged (and this
+  // capability should eventually be removed). Like `put`, this should not be
+  // called directly and should only be called by `run`, below.
+  template <typename T>
+  void put(T &t)
+  {
+    int type_id = TypeId<T>::type_id();
+    extern_state_.emplace(type_id, t);
+  }
+
+  // Will register the given type id and the name, which is resolvable via
+  // the `type_names_` map. This is called exclusively by `type_id` above.
+  static int add_type(std::string &&type_name)
+  {
+    int type_id = next_type_id_++;
+    type_names_[type_id] = std::move(type_name);
+    return type_id;
+  }
+
+  // Provide a short-hand to determine:
+  // - Whether the context contains an `ASTContext`, and,
+  // - Whether this `ASTContext` has error diagnostics.
+  //
+  // If both of the above are true, then false is returned. This is a
+  // convenience method provided for the `PassManager` below, as passes will
+  // stop when an error diagnostic has been generated.
+  bool ok() const;
+
+  static std::atomic<int> next_type_id_;
+  static std::unordered_map<int, std::string> type_names_;
+
+  std::unordered_map<int, std::unique_ptr<State>> state_;
+  std::unordered_map<int, std::reference_wrapper<State>> extern_state_;
+
+  friend class PassManager;
+  friend class Pass;
 };
 
-using PassFPtr = std::function<void(PassContext &)>;
+// State is the template class used for states that may be stored in a
+// PassContext. In order to consume or produce information in a given pass,
+// simply override this class with the desired type.
+//
+// For example:
+//
+//   class Variables : State<"variables"> {
+//     // my fields
+//   };
+//
+// The `Variables` class can then be stored in the pass context.
+template <PassContext::TypeName name>
+class State : public PassContext::State {
+public:
+  static std::string type_name()
+  {
+    return name.str();
+  }
+};
 
-// Base pass
+// Pass is an arbitrary pass.
+//
+// It is permitted to produce either no output or a single output, which may be
+// wrapped in `Result`, and may consume zero, one or multiple inputs.
+//
+// In general, this type should be constructed automatically, e.g.:
+//
+//    Pass CreateFooPass() {
+//      return Pass::create("mypass", [](MyInput &input) {
+//        return MyOutput();
+//      });
+//    }
+//
+// Pass functions must already return an `Result` instance.
 class Pass {
 public:
-  Pass() = delete;
-  Pass(std::string name, PassFPtr fn)
-      : fn_(std::move(fn)), name(std::move(name)) {};
+  Pass(const Pass &) = delete;
+  Pass(Pass &&) = default;
+  Pass &operator=(const Pass &) = delete;
+  Pass &operator=(Pass &&) = default;
 
-  virtual ~Pass() = default;
+  // create is the standard Pass constructor.
+  //
+  // The passed function may accept an arbitrary set of inputs. Note that the
+  // `type_identity_t` basically makes it so this template will always match,
```

Comment:
type_identity_t? I don't see any further references to it

Reply:
Yeah, this got rewritten a few times for sure. The `type_identity_t` was the original name for `TypeId` I think :) 

Anyways, I've simplified this documentation and fixed the below as well. Should now make sense.